### PR TITLE
Wrap-up changes for updated subfolders

### PR DESCRIPTION
**Pull request by artonge**
Add changes to support sharing and sub folders brought by the v2 encryption format.

- Add special handling for unsharing files fromE2EE folders, which conflicted with the deletion process
- Support wrapping up changes for subfolders
- Add counter logic to prevent overriding more recent metadata, and prevents rollback attacks
- Add signature logic to protect against metadata update from unauthorized parties 
- We need to support v1 and v2 encryption version, so we duplicated some code
- List of changes per files:
  - `MetaDataController`
  	- Add support for signature
  	- Improve locking logic for filedrop updates
  - `MetaDataStorage`
  	- Add support for signature and counter
  	- Keep track of touched folders
  - `LockingController`
  	- Add support for counter
  	- Support aborting changes during unlock
  	- Support finalizing changes in touch subfolders
  - `LockManager`
  	- Check counter when locking file
  	- Add possibility to assert locked by token in `isLocked`
  - `V1/MetaDataController`	
  	- Improve locking logic for filedrop updates
  - `MetadataStorageV1` // No changes
  - `V1/LockingController` // No changes
  - `LockStorageV1`
  	- Add possibility to assert locked by token in `isLocked`
  - `RollbackServiceV1` // No changes
  - `E2EEPublicShareTemplateProvider`
  	- Add encryption version and users public keys in initial state
  - Front files: adapt to new API requirements
- `ClientHasCapabilityMiddleware` --> Return an error to old clients when they access a v2.0 encrypted folder.

Fix:
- https://github.com/nextcloud/end_to_end_encryption/issues/438
- https://github.com/nextcloud/end_to_end_encryption/issues/437
- https://github.com/nextcloud/end_to_end_encryption/issues/479

### CODE REVIEW

File: src/services/filedrop.js [hunk 1]
```diff
@@ -20,6 +20,9 @@ import { EncryptedFile, encryptStringAsymmetric, encryptWithAES, getRandomAESKey
  * @property {string} encryptedTag - Encryption tag used to decrypt the 'encrypted' property.
  */
 
+/**
+ *
+ */
```
Comment (artonge) [inline src/services/filedrop.js:25]: To remove

File: lib/MetaDataStorage.php [hunk 2]
```diff
@@ -308,4 +313,34 @@ protected function getLegacyOwnerPath(string $userId, int $id):string {
...
+			fn (ISimpleFile $file) => (int)$file->getName(),
+			$this->getTokenFolder($token)->getDirectoryListing()
+		);
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function clearTouchedFolders(?string $token): void {
+		if ($token !== null) {
```
Comment (artonge) [inline lib/MetaDataStorage.php:331]: Why can token be null?

File: lib/Controller/LockingController.php [hunk 3]
```diff
@@ -152,16 +151,15 @@ public function unlockFolder(int $id, ?string $shareToken = null): DataResponse
...
-		try {
-			$this->metaDataStorage->saveIntermediateFile($ownerId, $id);
-		} catch (MissingMetaDataException $ex) {
-			if ($hadChanges) {
-				throw $ex;
-			}
+			$this->metaDataStorage->saveIntermediateFile($ownerId, $folderId);
 		}
 
+		$this->metaDataStorage->clearTouchedFolders($token);
```
Comment (artonge) [inline lib/Controller/LockingController.php:161]: Other places where we need to clear touched folders?

File: lib/Controller/MetaDataController.php [hunk 4]
```diff
@@ -161,8 +163,10 @@ public function updateMetaData(int $id, string $metaData): DataResponse {
 	 * @throws OCSBadRequestException
 	 */
 	public function deleteMetaData(int $id): DataResponse {
+		$e2eToken = $this->request->getParam('e2e-token');
```
Comment (artonge) [inline lib/Controller/MetaDataController.php:166]:
```suggestion
$e2eToken = $this->request->getHeader('e2e-token');
```

**Comment by artonge** [comment]
- Add `v2` routes in `routes.php`

- Copy-paste
    - Controllers
        - `lib/Controller/LockingController.php` -> `lib/Controller/v1/LockingController.php`
        - `lib/Controller/MetaDataController.php` -> `lib/Controller/v1/MetaDataController.php`
    - Services
        - `lib/MetaDataStorage.php` -> `lib/MetaDataStorageV1.php`
        - `lib/LockManager.php` -> `lib/LockManagerV1.php`
        - `lib/RollbackService.php` -> `lib/RollbackServiceV1.php`
        - `lib/IMetaDataStorage.php` -> `lib/IMetaDataStorageV1.php`
    - Unit tests
        - `tests/Unit/LockManagerTest.php` -> `lib/Unit/LockManagerV1Test.php`
        - `tests/Unit/MetaDataStorageTest.php` -> `lib/Unit/MetaDataStorageV1Test.php`
        - `tests/Unit/RollbackServiceTest.php` -> `lib/Unit/RollbackServiceV1Test.php`

- Support v1 and v2 in `filedrop.js`

File: lib/Controller/V1/MetaDataController.php [hunk 5]
```diff
@@ -130,6 +130,12 @@ public function setMetaData(int $id, string $metaData): DataResponse {
 	public function updateMetaData(int $id, string $metaData): DataResponse {
 		$e2eToken = $this->request->getParam('e2e-token');
 
+		// FIXME Temporary fix to handle both routes on single endpoint
+		if (empty($e2eToken)) {
+			$e2eToken = $this->request->getHeader('e2e-token');
+		}
+		// End
+
```
Comment (artonge) [inline lib/Controller/V1/MetaDataController.php:138]: To not forget to remove it

**Comment by artonge** [comment]
Filedrop need change:

- Add users' public keys to initial state data
- Create random symmetric key
- Use that key to encrypt dropped files
- Encrypt key with keys from the users array
- Upload an array with files info + array of encrypted symmetric keys for users + nonce + authTag

File: src/views/FileDrop.vue [hunk 6]
```diff
@@ -179,28 +169,19 @@ export default {
...
-				logger.debug('Unlocking the folder', { lockToken, shareToken: this.shareToken })
-				progresses
-					.filter(({ error }) => !error)
-					.forEach(progress => { progress.step = UploadStep.DONE })
+				await uploadFileDrop(this.encryptionVersion, this.folderId, fileDrops, this.shareToken)
 			} catch (exception) {
-				logger.error('Error while unlocking the folder', { exception })
-				showError(t('end_to_end_encryption', 'Error while unlocking the folder'))
+				logger.error('[FileDrop] Error while uploading metadata', { exception })
+				showError(this.t('end_to_end_encryption', 'Error while uploading metadata ({message})', { message: exception.response.data?.ocs?.meta?.message }))
```
[the remark this answers is not in the document]
Comment (artonge) [reply src/views/FileDrop.vue:177]: You're right

File: tests/Unit/RollbackServiceV1Test.php [hunk 7]
```diff
@@ -0,0 +1,240 @@
+<?php
```
Comment (Fenn-CS) [inline tests/Unit/RollbackServiceV1Test.php:1]: Not a big deal for these renamed files, but there is way to add them in git that would have made them show in the version control history as renamed and not newly added.

File: lib/Controller/V1/MetaDataController.php (hunk 5 above)
Comment (Fenn-CS) [reply lib/Controller/V1/MetaDataController.php]: @artonge are you still aware of this??/

File: src/services/filedrop.js [hunk 8]
```diff
@@ -103,10 +105,11 @@ export async function uploadFileDrop(folderId, fileDrop, lockToken, shareToken)
 		{
 			headers: {
 				'x-e2ee-supported': true,
-				'e2e-token': lockToken,
+				...(encryptionVersion === 2 ? { 'e2e-token': lockToken } : {}),
 			},
 			params: {
 				shareToken,
+				...(encryptionVersion === 1 ? { 'e2e-token': lockToken } : {}),
```
Comment (Fenn-CS) [inline src/services/filedrop.js]: Should this be version 2 instead? Is the `e2e-token` supported in v1?

```suggestion
...(encryptionVersion === 2 ? { 'e2e-token': lockToken } : {}),
```

Comment (artonge) [reply src/services/filedrop.js]: Yes, e2e token is the lock token. It was passed as a param in v1, and is now pased as a header in v2.

File: tests/Unit/RollbackServiceV1Test.php (hunk 7 above)
Comment (artonge) [reply tests/Unit/RollbackServiceV1Test.php:1]: How could I've done that?
In any case, the diff is shown for the v2 files, which is helpful.

**Review by Fenn-CS — APPROVED** [review]

File: tests/Unit/RollbackServiceV1Test.php (hunk 7 above)
Comment (Fenn-CS) [reply tests/Unit/RollbackServiceV1Test.php:1]: git mv OLD-FILENAME NEW-FILENAME

**Pull request merged**